gh-108590: Improve sqlite3 docs on encoding issues and how to handle those

erlend-aasland opened the pull request:
Document how to handle table columns with invalid Unicode sequences.


<!-- gh-issue-number: gh-108590 -->
* Issue: gh-108590
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--108699.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

erlend-aasland committed:
gh-108590: Add sqlite3 text factory howto

Document how to handle table columns with invalid Unicode sequences.

AlexWaygood commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
+Unicode sequences and other encodings.
+To work around this, you can use a custom :attr:`~Connection.text_factory`.
+
+Because of SQLites flexible typing, it is not uncommon to encounter table
```
```suggestion
Because of SQLite's flexible typing, it is not uncommon to encounter table
```

AlexWaygood commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
+
+Because of SQLites flexible typing, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing arbitrary data.
+Let's create a test database with an invalid Unicode sequence:
```
Maybe make clear that this isn't necessarily a _recommendation_, just a demonstration:

```suggestion
To demonstrate, let's create a test database with an invalid Unicode sequence:
```

AlexWaygood replied:
Maybe make clear that this isn't necessarily a _recommendation_, just a demonstration:

```suggestion
To demonstrate, let's create a test database with an invalid Unicode sequence:
```

AlexWaygood commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
+       INSERT INTO test VALUES(CAST(X'619F' AS TEXT));
+   """)
+
+To work with such databases, we can use the following trick,
```
"Trick" sounds a _tad_ colloquial here ;)

```suggestion
To work with such databases, we can use the following technique,
```

AlexWaygood commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
+   INSERT INTO "test" VALUES('a\udc9f');
+   COMMIT;
+
+Notice that in order to write the invalid Unicode sequence to a file,
```
```suggestion
Note that in order to write the invalid Unicode sequence to a file,
```

AlexWaygood commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
+   COMMIT;
+
+Notice that in order to write the invalid Unicode sequence to a file,
+you must also use ``errors="surrogateescape"`` as an argument to :func:`open`:
```
```suggestion
you will also have to use ``errors="surrogateescape"`` as an argument to :func:`open`:
```

erlend-aasland committed:
Apply suggestions from code review

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

serhiy-storchaka reviewed (commented):
`print()` will fail, because `sys.stdout.errors` is "strict" by default. It is not a new issue, you have the same issue when you print the result of `os.listdir()`, for example.

Other issue is that you cannot simply feed the result of `iterdump()` to `execute()`, because the Python interface only accept UTF-8 encodable string.

A minor problem is that using anything besides `str`, `bytes` or `bytearray` as `text_factory` slows down requests. So you only need to use this if necessary.

erlend-aasland committed:
Add note to iterdump(); explain why we use CAST

erlend-aasland committed:
Merge Alex's review

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Mention that execute() and friends only accept UTF-8 encoded strings

serhiy-storchaka commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,70 @@ With some adjustments, the above recipe can be adapted to use a
+   con = sqlite3.connect(":memory:")
+   con.executescript("""
+       CREATE TABLE test (data TEXT);
+       INSERT INTO test VALUES(CAST(X'619F' AS TEXT));
```
Are you from Norway? Did you use some historical character encodings different from ISO-8859-1?

I think that you can use more real-world example here. Some Norwegian geographic or personal name containing non-ASCII characters encoded in historical Norwegian character encoding. It is better to use something different from Latin1 and cp1252 because they are used sometimes as default fallback encoding, and we want to show that it is not the case here.

erlend-aasland replied:
> Are you from Norway?

I am.

> Did you use some historical character encodings different from ISO-8859-1?

Not that I can recall. IIRC, most webpages back in the 90s used ISO-8859-1.

> I think that you can use more real-world example here.

I agree; real-world examples are always better than contrived examples.

> Some Norwegian geographic or personal name containing non-ASCII characters encoded in historical Norwegian character encoding. It is better to use something different from Latin1 and cp1252 because they are used sometimes as default fallback encoding, and we want to show that it is not the case here.

Let's try and find something :)

erlend-aasland replied:
cc. @corona10 (as mentioned on the Discord chat)

ezio-melotti replied:
Using a different iso-8859-X (where `X != 1`) might be a good compromise, even though all iso-8859-X encoding can decode any byte sequence without decoding errors (even though the result might be unexpected).

> encoded in historical Norwegian character encoding

I'm not aware of any specific encoding other than `iso-8859-1`.
The Runic Unicode block comes to mind, but afaik that can only be encoded by UTF-* encodings :upside_down_face:), and wouldn't solve the problem at hand.

serhiy-storchaka replied:
@encukou proposed to use a Czech example with iso-8859-2, and I think a Czech example is perfect here.

erlend-aasland replied:
I hope the example of 3b9f620aacac66c20344473a218af5d75959736f is sufficient for now. Also note that we can cook up a better example in a later PR.

erlend-aasland committed:
Try to emphasize that strings with surrogate escapes cannot be passed to sqlite3 API

erlend-aasland committed:
Add seealso for the Unicode HOWTO

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -1157,6 +1157,13 @@ Connection objects
+      .. note::
+
+         If your database contains ``TEXT`` values with invalid Unicode
+         sequences, or encodings incompatible with UTF-8,
```
What does "invalid Unicode sequences" refer to?  The database should contain bytes, and it seems that these bytes are expected to be valid UTF-8.  There are bytes sequences that are invalid in UTF-8 (but might be valid for other encodings), but if the UTF-8 decoding succeeds, then the resulting sequence of Unicode characters should be ok (unless we are talking about surrogate pairs, some combining characters, or something else that I might be missing).

erlend-aasland replied:
We're talking about byte sequences that cannot be decoded to valid UTF-8. For example:

```python
>>> b"\x61\x9f".decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x9f in position 1: invalid start byte
```

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for correctly encoded UTF-8 text, but it will fail for invalid
+Unicode sequences and other encodings.
```
Here it should probably just talk about "valid UTF-8" or "text correctly encoded as UTF-8" (note that "UTF-8 text" doesn't really exist since UTF-8 is bytes, and you can't encode UTF-*).

erlend-aasland replied:
Good point :) I'll reword it tomorrow.

erlend-aasland replied:
Addressed in 3b9f620aacac66c20344473a218af5d75959736f

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
+using :class:`str`.
+This works well for correctly encoded UTF-8 text, but it will fail for invalid
+Unicode sequences and other encodings.
+To work around this, you can use a custom :attr:`~Connection.text_factory`.
```
Even though defaulting on UTF-8 is technically a limitation, I wouldn't say it's something that needs to be "worked around"
```suggestion
You can use a custom :attr:`~Connection.text_factory` to specify different encodings.
```

erlend-aasland replied:
Yeah, "work around" has a negative touch to it. Perhaps:

```suggestion
You can use a custom :attr:`~Connection.text_factory` to handle such cases.
```

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing arbitrary data.
+To demonstrate, let's create a test database with an invalid Unicode sequence.
```
Same thing, you want to create a field that contains a byte sequence that is not valid UTF-8 (or that can't be decoded as UTF-8).

erlend-aasland replied:
Yes, this is poorly worded. I'll suggest a more accurate wording tomorrow.

erlend-aasland replied:
How does 3b9f620aacac66c20344473a218af5d75959736f look?

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
+   for line in dump:
+       print(line)
+
+The loop above will print the offending line using Unicode surrogate escapes:
```
"offending" might be a bit harsh if the DB simply uses a different encoding for all the text.  If instead a non-UTF-8 byte sequence managed to sneak in among the rest of the UTF-8 then the term might ok (even though I would be more offended by the code that would have let that happen in the first place).

Maybe it would be useful to address these two cases separately?

erlend-aasland replied:
> "offending" might be a bit harsh if the DB simply uses a different encoding for all the text.

+1! 😄 

> Maybe it would be useful to address these two cases separately?

Perhaps we only need one of them; I think the case of a different encoding makes for the better example.

ezio-melotti replied:
Maybe this is something that we could revisit in a separate PR, but I have the feeling that the focus of this howto is to explain how to use a certain API (namely creating/using text factories) rather than solving a real world problem (i.e. how to deal with non-UTF-8 text fields).  IOW, I think it's unlikely that users will come to the docs looking for text factories (i.e. a solution they might not be aware of) -- they will probably come with a decoding error that they can't figure out (i.e. a problem).

The decoding error might have happened for different reasons depending on what exactly it's on those fields, and the way you deal with it also changes:
* for fields that simply use a different encoding, the solution is to use a text factory that decodes the text using the correct encoding;
* for fields that should be using UTF-8 but contain some ill-encoded byte sequence, the solution is to use `surrogateescape` and friends to escape/ignore the invalid bytes.

Since both problems are related to encodings and can be solved with text factories, you could cover both in the same howto, by explaining the "different encoding" scenario and then just adding a paragraph at the end to mention how to use `surrogateescape`.

erlend-aasland replied:
I reworded the how-to to focus on the problem (encodings) instead of the solution (text factories). I simplified the text (how-to's are not explanations) and tried to cover both cases. How does 3b9f620aacac66c20344473a218af5d75959736f look, Ezio?

serhiy-storchaka reviewed (commented):
I think we all agree that the current version has serious issues. No need to beat it more, it is already dead. Let @erlend-aasland to address our comments and prepare a new version. Then we will start a new round of bikeshedding.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Try to address reviews:

- Use a Latin 2 example; a Czech-English dictionary.
  We need the _values_ to be encoded in Latin 2,
  not the _column names_.

- Reword the title; the how-to is about the problem, not the means.

- Reword UTF-8 passages as requested by Ezio.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Fix note

serhiy-storchaka commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2596,65 @@ With some adjustments, the above recipe can be adapted to use a
+   Strings containing surrogate escapes and must be treated with care.
+   You cannot simply pass them back to SQLite,
+   for example using :meth:`~Cursor.execute`,
+   since the :mod:`!sqlite3` module APIs only accept UTF-8 encoded strings.
```
UTF-8 encodable strings.

erlend-aasland replied:
Why do you mean a `str` instance is best described as a "UTF-8 encodable string"? Can you elaborate why "encodable strings" is a better term?

serhiy-storchaka replied:
Because if it is UTF-8 encoded, it is no longer a `str` object. It is a bytes sequence, which can be represented in Python as `bytes`, bytearray`, etc. The sqlite module APIs do no support them.

Maybe "UTF-8 encodable string" does not sound well, I do not know, but "UTF-8 encoded string" is not correct here.

erlend-aasland replied:
See https://github.com/python/cpython/pull/108699#discussion_r1314662069 for suggestions.

serhiy-storchaka reviewed (commented):
It would be nice to document also `text_factory=bytes` option present in the old example. And `iterdumps()` needs special mentioning, because it does not work with `bytes`.

I do not think that it is needed to describe how to create an SQLite file with non-UTF-8 encoding, especially if it involves a third-party tool. It is not the problem which the user needs to solve, their problem is how to work with such files. Such code can go in tests.

erlend-aasland commented on Doc/library/sqlite3.rst:
```diff
@@ -1562,6 +1539,8 @@ Cursor objects
 
       Use :meth:`executescript` to execute multiple SQL statements.
 
+      :meth:`!execute` only accepts UTF-8 encoded strings.
```
@serhiy-storchaka, do you oppose this change?

erlend-aasland replied:
Never mind, I removed them; the PR now only focuses on the how-to.

serhiy-storchaka replied:
It is technically incorrect. I am not sure what wording would be better here and whether a special addition is needed at all, because for functions have the same requirement, but it is not documented, and in general you cannot know whether surrogate characters are acceptable until you try.

It is not actually clear to what the note refers. Is it only *sql* or *parameters*? Parameters can contain not only strings, but numbers, bytes and custom data. Strings must be UTF-8 encodable, but bytes can be not only UTF-8 encoded.

erlend-aasland replied:
*sql* is documented as being a `str` instance; not a `bytes` instance. There is no need to complicate the docs with encoding details here.

erlend-aasland commented on Doc/library/sqlite3.rst:
```diff
@@ -1610,6 +1589,8 @@ Cursor objects
          Starting with Python 3.14, :exc:`ProgrammingError` will
          be raised instead.
 
+      :meth:`!executemany` only accepts UTF-8 encoded strings.
```
@serhiy-storchaka, ditto

erlend-aasland commented on Doc/library/sqlite3.rst:
```diff
@@ -1635,6 +1616,7 @@ Cursor objects
              COMMIT;
          """)
 
+      :meth:`!executescript` only accepts UTF-8 encoded strings.
```
@serhiy-storchaka, ditto

erlend-aasland commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2596,65 @@ With some adjustments, the above recipe can be adapted to use a
+   Strings containing surrogate escapes and must be treated with care.
+   You cannot simply pass them back to SQLite,
+   for example using :meth:`~Cursor.execute`,
+   since the :mod:`!sqlite3` module APIs only accept UTF-8 encoded strings.
```
@serhiy-storchaka

Alt 1:

```suggestion
   Strings containing surrogate escapes and must be treated with care.
   You cannot simply pass them back to SQLite.
```

Alt 2:

```suggestion
   Strings containing surrogate escapes and must be treated with care.
   You cannot simply pass them back to SQLite;
   the :mod:`!sqlite3` module API does not understand
   surrogate escape codes.
```

Other suggestions?

erlend-aasland replied:
Alt 3: Just remove the note.

serhiy-storchaka replied:
Better ask @ezio-melotti.

CorvinM replied:
My 2 cents
Alt 4:
```suggestion
   Strings containing surrogate escapes must be treated with care as
   modules that attempt to encode them require support for surrogate
   escapes to handle them correctly.
   The :mod:`!sqlite3` module API does not support surrogate escapes
   so these strings cannot be passed back to the :mod:`!sqlite3` module API.
```

erlend-aasland replied:
```suggestion
   Strings containing surrogate escapes and must be treated with care.
   For example, you cannot pass them back to SQLite,
   since the :mod:`!sqlite3` module API does not support strings
   containing surrogate escape codes.
```

erlend-aasland committed:
Address review: don't need to show how to create the database

erlend-aasland committed:
Remove execute*() changes

erlend-aasland commented on Doc/library/sqlite3.rst:
```diff
@@ -1223,6 +1227,10 @@ Connection objects
 
       .. versionadded:: 3.7
 
+      .. seealso:
```
```suggestion
      .. seealso::
```

erlend-aasland committed:
Update Doc/library/sqlite3.rst

erlend-aasland committed:
Update Doc/library/sqlite3.rst

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for UTF-8 encoded text, but it will fail for other encodings
+and invalid UTF-8.
```
This is understandable and probably ok, but technically you might tweak it to say:
```suggestion
This works well for UTF-8 encoded text, but it might fail for other encodings
and invalid UTF-8.
```
Since e.g. ASCII-encoded text will also work (UTF-8 is ASCII-compatible), or even iso-8859-1-encoded text that contain only ASCII will fail.  IOW, it's not guaranteed that having text encoded with a different encoding will result in a failure.

erlend-aasland replied:
Thanks, good point.

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing non-UTF-8 encodings,
+or even arbitrary data.
```
```suggestion
Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
columns with the ``TEXT`` data type containing non-UTF-8 encodings,
or even arbitrary data.
```
Is this comma needed?

CAM-Gerlach replied:
Yeah, the comma implies the structure is:

> it is not uncommon to encounter table columns ((with the ``TEXT`` data type), (containing non-UTF-8 encodings), (or even arbitrary data)).

But it seems the intended meaning is instead:

> it is not uncommon to encounter table columns with the ``TEXT`` data type ((containing non-UTF-8 encodings), (or even arbitrary data)).

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing non-UTF-8 encodings,
+or even arbitrary data.
+To demonstrate, let's assume we've got a database with ISO-8859-2 (Latin 2)
```
```suggestion
To demonstrate, let's assume we have a database with ISO-8859-2 (Latin 2)
```

CAM-Gerlach replied:
Also, conventional spelling of "Latin-2" is with a hyphen:

```suggestion
To demonstrate, let's assume we have a database with ISO-8859-2 (Latin-2)
```

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+
+.. note::
+
+   Strings containing surrogate escapes and must be treated with care.
```
```suggestion
   Strings containing surrogate escapes must be treated with care.
```
Is that a stray `and` or did you want to add something else?

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+   Strings containing surrogate escapes and must be treated with care.
+   For example, you cannot pass them back to SQLite,
+   since the :mod:`!sqlite3` module API does not support strings
+   containing surrogate escape codes.
```
The wording of this sentence is a bit off.  You can talk about surrogates if you are talking about the characters themselves.  `surrogateescape` is an error handler that converts surrogates to Unicode characters in a different range, and you also have escape sequences (`'\uXXXX'`) that represent surrogates.

Does "pass them back" imply that you can get them out from the DB?

erlend-aasland replied:
Hm, yes this is certainly a bit off. The point we want to highlight (IMO) is that strings containing surrogate escape sequences will not be accepted by the `sqlite3` APIs. For example, `Cursor.execute` will not accept strings with surrogate escape sequences neither as the *sql* string, nor as a part of the *parameters* sequence.

erlend-aasland replied:
Let's unblock this PR by removing this note; we can add a note in a follow-up PR.

CAM-Gerlach commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+encoded text, for example a table of Czech-English dictionary entries.
+Assuming we now have a :class:`Connection` instance :py:data:`!con`
+connected to this database,
+we can decode the Latin 2 encoded text using this :attr:`~Connection.text_factory`:
```
```suggestion
we can decode the Latin-2 encoded text using this :attr:`~Connection.text_factory`:
```

Minor tweak

CAM-Gerlach commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
```
```suggestion
By default, :mod:`!sqlite3` uses :class:`str` to adapt SQLite values
with the ``TEXT`` data type.
```

I find this phrasing clearer, less ambiguous and easier to follow

erlend-aasland replied:
Thanks, that reads better.

CAM-Gerlach reviewed (commented):
Looks very good now, thanks! Just a few small suggestions.

erlend-aasland committed:
Apply suggestions from CAM and Ezio

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>
Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
+   Strings containing surrogate escapes must be treated with care.
+   For example, you cannot pass them back to SQLite,
+   since the :mod:`!sqlite3` module API does not support strings
+   containing surrogate escape codes.
```
```suggestion
   The :mod:`!sqlite3` module API does not support strings
   containing surrogates.
```
I would simply say this then.  This will make the users aware of this limitation without going in too much details, so that they will remove the surrogates and try again.  If there are situations where surrogates are allowed (e.g. can I have them in a db and read them?) then it might be worth to specify that.

erlend-aasland replied:
The example just above this text shows how to deal with strings w/surrogates in the DB. I think your proposal works swell. Keep it simple 🪄

erlend-aasland replied:
Alternatively, one could remove this warning and mention this limitation explicitly for each API.

erlend-aasland committed:
Keep it simple

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

serhiy-storchaka reviewed (commented):
This change removes any mention of `bytes`. It makes the documentation less informative.

It also does not resolve the original issue, in particularly, it does not document that `iterdump()` is not compatible with `text_factory=bytes`. And it does not say what to do with the result of `iterdump()` with custom `text_factory`.
